python/pyspark/sql/variant_utils.py

```diff
@@ -0,0 +1,385 @@
+    # names in one object are not allowed.
+    # We use 5 bits in the type info to specify the integer type of the object header: it should
+    # be 0_b4_b3b2_b1b0 (MSB is 0), where:
+    # - b4 specifies the type of size. When it is 0/1, `size` is a little-endian 1/4-byte
+    # unsigned integer.
+    # - b3b2/b1b0 specifies the integer type of id and offset. When the 2 bits are  0/1/2, the
+    # list contains 1/2/3-byte little-endian unsigned integers.
+    OBJECT = 2
+    # Array value. The content contains a size, a list of field offsets, and the actual element
+    # data. It is similar to an object without the id list. The length of the offset list
+    # is `size + 1`, where the last offset represent the total size of the element data.
+    # Its type info should be: 000_b2_b1b0:
+    # - b2 specifies the type of size.
+    # - b1b0 specifies the integer type of offset.
+    ARRAY = 3
+
+    # Below is all possible type info values for `PRIMITIVE`.
+    # JSON Null value. Empty content.
+    NULL = 0
+    # True value. Empty content.
+    TRUE = 1
+    # False value. Empty content.
+    FALSE = 2
+    # 1-byte little-endian signed integer.
+    INT1 = 3
+    # 2-byte little-endian signed integer.
+    INT2 = 4
+    # 4-byte little-endian signed integer.
+    INT4 = 5
+    # 4-byte little-endian signed integer.
+    INT8 = 6
+    # 8-byte IEEE double.
+    DOUBLE = 7
+    # 4-byte decimal. Content is 1-byte scale + 4-byte little-endian signed integer.
+    DECIMAL4 = 8
+    # 8-byte decimal. Content is 1-byte scale + 8-byte little-endian signed integer.
+    DECIMAL8 = 9
+    # 16-byte decimal. Content is 1-byte scale + 16-byte little-endian signed integer.
+    DECIMAL16 = 10
+    # Long string value. The content is (4-byte little-endian unsigned integer representing the
+    # string size) + (size bytes of string content).
+    LONG_STR = 16
+
+    U32_SIZE = 4
+
+    @classmethod
+    def to_json(cls, value: bytes, metadata: bytes) -> str:
+        """
+        Convert the VariantVal to a JSON string.
+        :return: JSON string
+        """
+        return cls._to_json(value, metadata, 0)
+
+    @classmethod
+    def to_python(cls, value: bytes, metadata: bytes) -> str:
+        """
+        Convert the VariantVal to a nested Python object of Python data types.
+        :return: Python representation of the Variant nested structure
+        """
+        return cls._to_python(value, metadata, 0)
+
+    @classmethod
+    def _read_long(cls, data: bytes, pos: int, num_bytes: int, signed: bool) -> int:
+        cls._check_index(pos, len(data))
+        cls._check_index(pos + num_bytes - 1, len(data))
+        return int.from_bytes(data[pos : pos + num_bytes], byteorder="little", signed=signed)
+
+    @classmethod
+    def _check_index(cls, pos: int, length: int) -> None:
+        if pos < 0 or pos >= length:
+            raise Exception("Malformed Variant")
```

Comment:
Such exceptions failed PySpark custom errors check, please see CI errors.

Reply:
Thanks! I updated to reuse an existing exception class.